File: advent_of_code_2025/day01/Main.hs
module Main where

import Control.Applicative ((<|>))
import Data.Attoparsec.Text
import Data.Foldable (traverse_)
import Data.Functor (($>))
import Data.Text (Text)
import Data.Text.IO qualified as T

parse_ :: Text -> [Int]
parse_ = either (error "Bad parse") id . parseOnly (line `sepBy1'` char '\n')
  where
    line = ((char 'L' $> negate) <|> (char 'R' $> id)) <*> decimal

part1 :: [Int] -> Int
part1 = length . filter (== 0) . scanl (\a b -> (a + b) `mod` 100) 50

part2 :: [Int] -> Int
part2 = fst . foldl' f (0, 50)
  where
    f (n, p) i
        | i == 0 = (n, p)
        | otherwise =
            let
                n' = n + if p' == 0 then 1 else 0
                p' = (p + s) `mod` 100
                s = signum i
             in
                f (n', p') (i - s)

main :: IO ()
main = do
    input <- parse_ <$> T.readFile "input.txt"
    traverse_ (print . ($ input)) [part1, part2]

File: advent_of_code_2025/day01/src/main.rs
fn parse(s: &str) -> Vec<i16> {
    s.trim()
        .lines()
        .map(|line| {
            let (x, y) = line.split_at(1);
            (if x == "L" { -1 } else { 1 }) * y.parse::<i16>().expect("val")
        })
        .collect()
}

fn part_1(input: &[i16]) -> usize {
    input
        .iter()
        .scan(50, |a, b| {
            *a += *b;
            *a %= 100;
            Some(*a)
        })
        .filter(|&n| n == 0)
        .count()
}

fn part_2(input: &[i16]) -> usize {
    input
        .iter()
        .fold((0, 50), |(mut n, mut p), i| {
            let s = i.signum();
            for _ in 0..(s * i) {
                p += s;
                p %= 100;
                n += usize::from(p == 0);
            }
            (n, p)
        })
        .0
}

fn main() {
    let input = parse(&include_str!("../input.txt"));
    println!("{}", part_1(&input));
    println!("{}", part_2(&input));
}
